File: infrastructure/services/connection-service/routes/connections.js
const express = require('express');
const router = express.Router();
const { v4: uuidv4 } = require('uuid');
const { getPool } = require('../config/db');
const { publishEvent } = require('../config/kafka');

const PROFILE_SERVICE_URL = process.env.PROFILE_SERVICE_URL || 'http://profile-service:8000';

async function incrementConnectionsCount(memberId) {
  try {
    const res = await fetch(`${PROFILE_SERVICE_URL}/members/${memberId}/increment-connections`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
    });
    if (!res.ok) {
      console.error(`increment-connections failed for ${memberId}: ${res.status}`);
    }
  } catch (e) {
    console.error(`increment-connections error for ${memberId}:`, e.message);
  }
}

// POST /api/connections/request — Send a connection request
router.post('/request', async (req, res) => {
  try {
    const { requester_id, receiver_id } = req.body;

    if (!requester_id || !receiver_id) {
      return res.status(400).json({
        success: false,
        message: 'requester_id and receiver_id are required'
      });
    }

    if (requester_id === receiver_id) {
      return res.status(400).json({
        success: false,
        message: 'Cannot send connection request to yourself'
      });
    }

    const pool = getPool();
    const connection_id = 'CON' + uuidv4().replace(/-/g, '').substring(0, 8).toUpperCase();

    // Check if a connection already exists in either direction
    const [existing] = await pool.execute(
      `SELECT * FROM connections
       WHERE (requester_id = ? AND receiver_id = ?)
          OR (requester_id = ? AND receiver_id = ?)`,
      [requester_id, receiver_id, receiver_id, requester_id]
    );

    if (existing.length > 0) {
      const conn = existing[0];
      if (conn.status === 'accepted') {
        return res.status(409).json({
          success: false,
          message: 'You are already connected with this user'
        });
      }
      if (conn.status === 'pending') {
        return res.status(409).json({
          success: false,
          message: 'A connection request already exists between these users',
          connection_id: conn.connection_id,
          status: conn.status
        });
      }
      // If previously rejected, allow a new request by updating the existing row
      if (conn.status === 'rejected') {
        await pool.execute(
          `UPDATE connections SET status = 'pending', requester_id = ?, receiver_id = ?,
           connection_id = ?, updated_at = NOW() WHERE id = ?`,
          [requester_id, receiver_id, connection_id, conn.id]
        );

        await publishEvent('connection.requested', requester_id, 'connection', connection_id, {
          connection_id,
          requester_id,
          receiver_id,
          status: 'pending'
        });

        return res.status(200).json({
          success: true,
          message: 'Connection request re-sent successfully',
          data: { connection_id, requester_id, receiver_id, status: 'pending' }
        });
      }
    }

    // Insert new connection request
    await pool.execute(
      `INSERT INTO connections (connection_id, requester_id, receiver_id, status)
       VALUES (?, ?, ?, 'pending')`,
      [connection_id, requester_id, receiver_id]
    );

    // Publish Kafka event: connection.requested
    await publishEvent('connection.requested', requester_id, 'connection', connection_id, {
      connection_id,
      requester_id,
      receiver_id,
      status: 'pending'
    });

    res.status(201).json({
      success: true,
      message: 'Connection request sent successfully',
      data: { connection_id, requester_id, receiver_id, status: 'pending' }
    });
  } catch (error) {
    // Handle duplicate entry (MySQL unique constraint violation)
    if (error.code === 'ER_DUP_ENTRY') {
      return res.status(409).json({
        success: false,
        message: 'Duplicate connection request'
      });
    }
    console.error('Error sending connection request:', error);
    res.status(500).json({ success: false, message: 'Internal server error' });
  }
});

// POST /api/connections/accept — Accept a pending connection request
router.post('/accept', async (req, res) => {
  try {
    const { connection_id } = req.body;

    if (!connection_id) {
      return res.status(400).json({ success: false, message: 'connection_id is required' });
    }

    const pool = getPool();
    const [rows] = await pool.execute(
      'SELECT * FROM connections WHERE connection_id = ?',
      [connection_id]
    );

    if (rows.length === 0) {
      return res.status(404).json({ success: false, message: 'Connection request not found' });
    }

    const conn = rows[0];

    if (conn.status === 'accepted') {
      return res.status(409).json({ success: false, message: 'Connection already accepted' });
    }

    if (conn.status !== 'pending') {
      return res.status(400).json({
        success: false,
        message: `Cannot accept a connection with status: ${conn.status}`
      });
    }

    await pool.execute(
      "UPDATE connections SET status = 'accepted', updated_at = NOW() WHERE connection_id = ?",
      [connection_id]
    );

    await Promise.all([
      incrementConnectionsCount(conn.requester_id),
      incrementConnectionsCount(conn.receiver_id),
    ]);

    // Publish Kafka event: connection.accepted
    try {
      await publishEvent('connection.accepted', conn.receiver_id, 'connection', connection_id, {
        connection_id,
        requester_id: conn.requester_id,
        receiver_id: conn.receiver_id,
        status: 'accepted'
      });
    } catch (kafkaErr) {
      console.error('connection.accepted Kafka publish failed:', kafkaErr.message);
    }

    res.status(200).json({
      success: true,
      message: 'Connection accepted successfully',
      data: {
        connection_id,
        requester_id: conn.requester_id,
        receiver_id: conn.receiver_id,
        status: 'accepted'
      }
    });
  } catch (error) {
    console.error('Error accepting connection:', error);
    res.status(500).json({ success: false, message: 'Internal server error' });
  }
});

// POST /api/connections/reject — Reject a pending connection request
router.post('/reject', async (req, res) => {
  try {
    const { connection_id } = req.body;

    if (!connection_id) {
      return res.status(400).json({ success: false, message: 'connection_id is required' });
    }

    const pool = getPool();
    const [rows] = await pool.execute(
      'SELECT * FROM connections WHERE connection_id = ?',
      [connection_id]
    );

    if (rows.length === 0) {
      return res.status(404).json({ success: false, message: 'Connection request not found' });
    }

    if (rows[0].status !== 'pending') {
      return res.status(400).json({
        success: false,
        message: `Cannot reject a connection with status: ${rows[0].status}`
      });
    }

    await pool.execute(
      "UPDATE connections SET status = 'rejected', updated_at = NOW() WHERE connection_id = ?",
      [connection_id]
    );

    res.status(200).json({
      success: true,
      message: 'Connection request rejected',
      data: {
        connection_id,
        requester_id: rows[0].requester_id,
        receiver_id: rows[0].receiver_id,
        status: 'rejected'
      }
    });
  } catch (error) {
    console.error('Error rejecting connection:', error);
    res.status(500).json({ success: false, message: 'Internal server error' });
  }
});

// POST /api/connections/list — List all connections for a user
router.post('/list', async (req, res) => {
  try {
    const { user_id, status } = req.body;

    if (!user_id) {
      return res.status(400).json({ success: false, message: 'user_id is required' });
    }

    const pool = getPool();
    let query = `SELECT * FROM connections WHERE (requester_id = ? OR receiver_id = ?)`;
    let params = [user_id, user_id];

    if (status) {
      query += ` AND status = ?`;
      params.push(status);
    }

    query += ` ORDER BY updated_at DESC`;

    const [rows] = await pool.execute(query, params);

    // Separate into sent, received, and accepted
    const connections = rows.map(row => ({
      connection_id: row.connection_id,
      requester_id: row.requester_id,
      receiver_id: row.receiver_id,
      status: row.status,
      direction: row.requester_id === user_id ? 'sent' : 'received',
      connected_user_id: row.requester_id === user_id ? row.receiver_id : row.requester_id,
      created_at: row.created_at,
      updated_at: row.updated_at
    }));

    res.status(200).json({
      success: true,
      user_id,
      total_connections: connections.length,
      connections
    });
  } catch (error) {
    console.error('Error listing connections:', error);
    res.status(500).json({ success: false, message: 'Internal server error' });
  }
});

// POST /api/connections/mutual — Get mutual connections between two users (extra credit)
router.post('/mutual', async (req, res) => {
  try {
    const { user_id_1, user_id_2 } = req.body;

    if (!user_id_1 || !user_id_2) {
      return res.status(400).json({
        success: false,
        message: 'user_id_1 and user_id_2 are required'
      });
    }

    const pool = getPool();

    // Find all accepted connections for user 1
    const [user1Connections] = await pool.execute(
      `SELECT CASE
         WHEN requester_id = ? THEN receiver_id
         ELSE requester_id
       END AS connected_user
       FROM connections
       WHERE (requester_id = ? OR receiver_id = ?) AND status = 'accepted'`,
      [user_id_1, user_id_1, user_id_1]
    );

    // Find all accepted connections for user 2
    const [user2Connections] = await pool.execute(
      `SELECT CASE
         WHEN requester_id = ? THEN receiver_id
         ELSE requester_id
       END AS connected_user
       FROM connections
       WHERE (requester_id = ? OR receiver_id = ?) AND status = 'accepted'`,
      [user_id_2, user_id_2, user_id_2]
    );

    const user1Set = new Set(user1Connections.map(r => r.connected_user));
    const user2Set = new Set(user2Connections.map(r => r.connected_user));

    // Mutual = intersection of both sets
    const mutual = [...user1Set].filter(id => user2Set.has(id));

    res.status(200).json({
      success: true,
      user_id_1,
      user_id_2,
      mutual_count: mutual.length,
      mutual_connections: mutual
    });
  } catch (error) {
    console.error('Error finding mutual connections:', error);
    res.status(500).json({ success: false, message: 'Internal server error' });
  }
});

module.exports = router;

File: services/connection-service/backend/app/api/connections.py
import uuid
from datetime import datetime, timezone
from typing import Any

import httpx
from fastapi import APIRouter, Body, Depends
from fastapi.responses import JSONResponse
from sqlalchemy import text
from sqlalchemy.orm import Session

from app.core.config import settings
from app.db.session import get_db
from app.kafka_prod import publish_event

router = APIRouter()


def _increment_profile_connections(member_id: str) -> None:
    primary = settings.profile_service_url.rstrip("/")
    bases = [primary]
    fallback = "http://host.docker.internal:8002"
    if fallback.rstrip("/") != primary:
        bases.append(fallback)
    for base in bases:
        url = f"{base}/api/members/{member_id}/increment-connections"
        try:
            with httpx.Client(timeout=10.0) as client:
                r = client.post(url)
                if r.status_code != 200:
                    print(f"increment-connections HTTP {r.status_code} for {member_id} @ {base}: {r.text[:300]}")
                    continue
                body = r.json()
                if body.get("success"):
                    return
                print(f"increment-connections profile refused for {member_id} @ {base}: {body}")
        except httpx.RequestError as e:
            print(f"increment-connections unreachable {url}: {e}")
            continue
        except Exception as e:
            print(f"increment-connections error for {member_id} @ {base}: {e}")


def _new_connection_id() -> str:
    return "CON" + uuid.uuid4().hex[:8].upper()


@router.post("/request")
def request_connection(body: dict[str, Any] = Body(default_factory=dict), db: Session = Depends(get_db)):
    requester_id = body.get("requester_id")
    receiver_id = body.get("receiver_id")
    if not requester_id or not receiver_id:
        return {"success": False, "message": "requester_id and receiver_id are required"}
    if requester_id == receiver_id:
        return {"success": False, "message": "Cannot send connection request to yourself"}

    rows = db.execute(
        text(
            """SELECT * FROM connections
               WHERE (requester_id = :a AND receiver_id = :b)
                  OR (requester_id = :b AND receiver_id = :a)"""
        ),
        {"a": requester_id, "b": receiver_id},
    ).fetchall()

    connection_id = _new_connection_id()

    if rows:
        conn_row = dict(rows[0]._mapping)
        st = conn_row["status"]
        if st == "accepted":
            return JSONResponse(
                status_code=409,
                content={"success": False, "message": "You are already connected with this user"},
            )
        if st == "pending":
            return JSONResponse(
                status_code=409,
                content={
                    "success": False,
                    "message": "A connection request already exists between these users",
                    "connection_id": conn_row["connection_id"],
                    "status": st,
                },
            )
        if st == "rejected":
            db.execute(
                text(
                    """UPDATE connections SET status = 'pending', requester_id = :rid, receiver_id = :rcv,
                       connection_id = :cid, updated_at = NOW() WHERE id = :id"""
                ),
                {
                    "rid": requester_id,
                    "rcv": receiver_id,
                    "cid": connection_id,
                    "id": conn_row["id"],
                },
            )
            db.commit()
            publish_event(
                "connection.requested",
                requester_id,
                "connection",
                connection_id,
                {
                    "connection_id": connection_id,
                    "requester_id": requester_id,
                    "receiver_id": receiver_id,
                    "status": "pending",
                },
            )
            return JSONResponse(
                status_code=201,
                content={
                    "success": True,
                    "message": "Connection request re-sent successfully",
                    "data": {
                        "connection_id": connection_id,
                        "requester_id": requester_id,
                        "receiver_id": receiver_id,
                        "status": "pending",
                    },
                },
            )

    try:
        db.execute(
            text(
                """INSERT INTO connections (connection_id, requester_id, receiver_id, status)
                   VALUES (:cid, :rid, :rcv, 'pending')"""
            ),
            {"cid": connection_id, "rid": requester_id, "rcv": receiver_id},
        )
        db.commit()
    except Exception as e:
        db.rollback()
        err = str(e).lower()
        if "duplicate" in err or "1062" in err:
            return JSONResponse(
                status_code=409,
                content={"success": False, "message": "Duplicate connection request"},
            )
        raise

    publish_event(
        "connection.requested",
        requester_id,
        "connection",
        connection_id,
        {
            "connection_id": connection_id,
            "requester_id": requester_id,
            "receiver_id": receiver_id,
            "status": "pending",
        },
    )
    return JSONResponse(
        status_code=201,
        content={
            "success": True,
            "message": "Connection request sent successfully",
            "data": {
                "connection_id": connection_id,
                "requester_id": requester_id,
                "receiver_id": receiver_id,
                "status": "pending",
            },
        },
    )


@router.post("/accept")
def accept(body: dict = Body(default_factory=dict), db: Session = Depends(get_db)):
    connection_id = body.get("connection_id")
    if not connection_id:
        return {"success": False, "message": "connection_id is required"}

    rows = db.execute(
        text("SELECT * FROM connections WHERE connection_id = :c"),
        {"c": connection_id},
    ).fetchall()
    if not rows:
        return {"success": False, "message": "Connection request not found"}
    conn_row = dict(rows[0]._mapping)
    if conn_row["status"] == "accepted":
        return {"success": False, "message": "Connection already accepted"}
    if conn_row["status"] != "pending":
        return {"success": False, "message": f"Cannot accept a connection with status: {conn_row['status']}"}

    db.execute(
        text("UPDATE connections SET status = 'accepted', updated_at = NOW() WHERE connection_id = :c"),
        {"c": connection_id},
    )
    db.commit()
    _increment_profile_connections(conn_row["requester_id"])
    _increment_profile_connections(conn_row["receiver_id"])
    try:
        publish_event(
            "connection.accepted",
            conn_row["receiver_id"],
            "connection",
            connection_id,
            {
                "connection_id": connection_id,
                "requester_id": conn_row["requester_id"],
                "receiver_id": conn_row["receiver_id"],
                "status": "accepted",
            },
        )
    except Exception as e:
        print(f"connection.accepted Kafka publish failed: {e}")

    return {
        "success": True,
        "message": "Connection accepted successfully",
        "data": {
            "connection_id": connection_id,
            "requester_id": conn_row["requester_id"],
            "receiver_id": conn_row["receiver_id"],
            "status": "accepted",
        },
    }


@router.post("/reject")
def reject(body: dict = Body(default_factory=dict), db: Session = Depends(get_db)):
    connection_id = body.get("connection_id")
    if not connection_id:
        return {"success": False, "message": "connection_id is required"}
    rows = db.execute(
        text("SELECT * FROM connections WHERE connection_id = :c"),
        {"c": connection_id},
    ).fetchall()
    if not rows:
        return {"success": False, "message": "Connection request not found"}
    conn_row = dict(rows[0]._mapping)
    if conn_row["status"] != "pending":
        return {"success": False, "message": f"Cannot reject a connection with status: {conn_row['status']}"}
    db.execute(
        text("UPDATE connections SET status = 'rejected', updated_at = NOW() WHERE connection_id = :c"),
        {"c": connection_id},
    )
    db.commit()
    return {
        "success": True,
        "message": "Connection request rejected",
        "data": {
            "connection_id": connection_id,
            "requester_id": conn_row["requester_id"],
            "receiver_id": conn_row["receiver_id"],
            "status": "rejected",
        },
    }


@router.post("/list")
def list_connections(body: dict = Body(default_factory=dict), db: Session = Depends(get_db)):
    user_id = body.get("user_id")
    status_f = body.get("status")
    if not user_id:
        return {"success": False, "message": "user_id is required"}
    q = "SELECT * FROM connections WHERE (requester_id = :u OR receiver_id = :u)"
    params: dict[str, Any] = {"u": user_id}
    if status_f:
        q += " AND status = :st"
        params["st"] = status_f
    q += " ORDER BY updated_at DESC"
    rows = db.execute(text(q), params).fetchall()
    connections = []
    for row in rows:
        r = dict(row._mapping)
        uid = user_id
        connections.append(
            {
                "connection_id": r["connection_id"],
                "requester_id": r["requester_id"],
                "receiver_id": r["receiver_id"],
                "status": r["status"],
                "direction": "sent" if r["requester_id"] == uid else "received",
                "connected_user_id": r["receiver_id"] if r["requester_id"] == uid else r["requester_id"],
                "created_at": str(r["created_at"]),
                "updated_at": str(r["updated_at"]),
            }
        )
    return {"success": True, "user_id": user_id, "total_connections": len(connections), "connections": connections}


@router.post("/mutual")
def mutual(body: dict = Body(default_factory=dict), db: Session = Depends(get_db)):
    u1 = body.get("user_id_1")
    u2 = body.get("user_id_2")
    if not u1 or not u2:
        return {"success": False, "message": "user_id_1 and user_id_2 are required"}

    r1 = db.execute(
        text(
            """SELECT CASE WHEN requester_id = :u THEN receiver_id ELSE requester_id END AS connected_user
               FROM connections WHERE (requester_id = :u OR receiver_id = :u) AND status = 'accepted'"""
        ),
        {"u": u1},
    ).fetchall()
    r2 = db.execute(
        text(
            """SELECT CASE WHEN requester_id = :u THEN receiver_id ELSE requester_id END AS connected_user
               FROM connections WHERE (requester_id = :u OR receiver_id = :u) AND status = 'accepted'"""
        ),
        {"u": u2},
    ).fetchall()
    s1 = {dict(x._mapping)["connected_user"] for x in r1}
    s2 = {dict(x._mapping)["connected_user"] for x in r2}
    mutual_ids = list(s1 & s2)
    return {"success": True, "user_id_1": u1, "user_id_2": u2, "mutual_count": len(mutual_ids), "mutual_connections": mutual_ids}


@router.get("/health")
def health():
    ts = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    return {"status": "ok", "service": "connection-service", "timestamp": ts}
